runtime/compiler/runtime/J9CodeCacheManager.cpp

```diff
@@ -355,18 +359,46 @@ J9::CodeCacheManager::allocateCodeCacheSegment(size_t segmentSize,
 
    void *defaultEndAddress = vmemParams.endAddress;
 
+   uintptr_t someJitLibraryAddress = self()->getSomeJitLibraryAddress();
+
+#if defined(TR_TARGET_POWER)
+   size_t alignment = 64 * 1024;
+#elif (defined(LINUX) && defined(TR_TARGET_X86))
+   size_t alignment = 2 * 1024 * 1024;
+#elif (defined(TR_TARGET_S390))
+   size_t alignment = 1024 * 1024;
+#else
+   size_t alignment = pageSizes[0];
+#endif
+   if (largeCodePageSize > 0)
+      alignment = largeCodePageSize;
+
    if (preferredStartAddress)
       {
-   // J9PORT_VMEM_ADDRESS_HINT is only supported for default pages and linux
+      vmemParams.options |= J9PORT_VMEM_STRICT_ADDRESS; // if we cannot allocate a block in the desired range, return NULL
+      vmemParams.alignmentInBytes = alignment;
 #if defined(LINUX)
-      if (largeCodePageSize == 0)
+      vmemParams.options |= J9PORT_VMEM_ALLOC_QUICK; // Use smaps for guidance
+      // If jit library address is greater than MAX_DISTANCE (~2GB),
+      // then we search in the addresses before the jit library address.
+      // If jit library address is less than MAX_DISTANCE, (which means we
+      // do not have the full MAX_DISTANCE range before the jit library),
+      // then we search in the addresses after the jit library address
+      if (someJitLibraryAddress > MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE)
          {
-         vmemParams.options |= J9PORT_VMEM_ADDRESS_HINT;
+         // Align the startAddress to page boundary
+         vmemParams.startAddress = (void *)OMR::align((size_t)(someJitLibraryAddress - MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE), alignment);
+         vmemParams.endAddress = (void *)(((UDATA)preferredStartAddress) + SAFE_DISTANCE_REPOSITORY_JITLIBRARY);
          }
+      else
+         {
+         vmemParams.startAddress = (void *)OMR::align((size_t)(someJitLibraryAddress + SAFE_DISTANCE_REPOSITORY_JITLIBRARY), alignment);
+         vmemParams.endAddress = (void *)(someJitLibraryAddress + MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE - segmentSize);
+         }
+#else // WIN64
```

Comment:
Isn't this everything other than Linux, so not _only_ WIN64.

Reply:
A preferred address is used only for x86-64, so practically the "else" is used only for Win64. I could remove the comment

Reply:
Well the `#if` part of this if/else section is
```
#if defined(LINUX)
```
so wouldn't the else be anything other than Linux, and not just WIN64? I guess that's why I wasn't sure why the else is only going to be executed for WIN64.